File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        maxCurr = nums[0]
        maxAns = nums[0]

        total = nums[0]

        minCurr = nums[0]
        minAns = nums[0]

        for i in range(1,len(nums)):

            total += nums[i]

            maxCurr = max(nums[i], maxCurr + nums[i])
            maxAns = max(maxAns,maxCurr)

            minCurr = min(nums[i],minCurr + nums[i])
            minAns = min(minCurr, minAns)

        
        if maxAns < 0:
            return maxAns

        return max(maxAns,total - minAns)
        

File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) 
    {
        int currMax = nums[0];
        int maxAns = nums[0];

        int total = nums[0];

        int currMin = nums[0];
        int minAns = nums[0];

        for(int i=1 ; i < nums.size() ;i++)
        {
            currMax = max(nums[i],currMax + nums[i]);
            maxAns = max(maxAns,currMax);

            total += nums[i];

            currMin = min(nums[i],currMin + nums[i]);
            minAns = min(minAns,currMin);
        }

        if ( maxAns < 0) return maxAns;

        return max(maxAns, total - minAns);
        
    }
};